File: src/main/java/io/github/mg138/tsbook/utils/config/item/ItemSetting.java
package io.github.mg138.tsbook.utils.config.item;

import io.github.mg138.tsbook.Book;
import org.bukkit.Material;
import org.bukkit.configuration.file.YamlConfiguration;

import java.util.List;

public class ItemSetting {
    public final String ID;
    public final String ITEM_TYPE;
    public final Material MATERIAL;
    public final Integer MODEL;
    public final String NAME;
    public final List<String> LORE;

    public ItemSetting(YamlConfiguration setting) {
        this(
                setting.getString("ID"),
                setting.getString("ITEM_TYPE"),
                Material.valueOf(setting.getString("MATERIAL")),
                setting.getInt("MODEL"),
                Book.Companion.getCfg().translate.translateString(setting.getString("NAME")),
                Book.Companion.getCfg().translate.translateByList(setting.getStringList("LORE"))
        );
    }

    public ItemSetting(ItemSetting setting) {
        this(
                setting.ID,
                setting.ITEM_TYPE,
                setting.MATERIAL,
                setting.MODEL,
                setting.NAME,
                setting.LORE
        );
    }

    public ItemSetting(String id, String item_type, Material material, Integer model, String name, List<String> lore) {
        ID = id;
        ITEM_TYPE = item_type;
        MATERIAL = material;
        MODEL = model;
        NAME = name;
        LORE = lore;
    }
}

File: src/main/kotlin/io/github/mg138/tsbook/setting/item/element/ItemSetting.kt
package io.github.mg138.tsbook.setting.item.element

import io.github.mg138.tsbook.item.attribute.Identifier
import io.github.mg138.tsbook.item.attribute.ItemType
import io.github.mg138.tsbook.setting.util.translate.TranslateUtil
import org.bukkit.Material
import org.bukkit.configuration.file.YamlConfiguration

open class ItemSetting(
    val id: Identifier,
    val itemType: ItemType,
    val material: Material,
    val model: Int = 0,
    val name: String,
    val lore: List<String>
) {
    init {
        if (material == Material.AIR) throw IllegalArgumentException("Material cannot be air! (ID: $id)")
    }

    constructor(setting: ItemSetting) : this(
        setting.id,
        setting.itemType,
        setting.material,
        setting.model,
        setting.name,
        setting.lore
    )

    constructor(setting: YamlConfiguration, key: String) : this(
        setting.getString("ID")?.let {
            Identifier(key, it)
        } ?: throw IllegalArgumentException("Setting doesn't contain an ID."),
        setting.getString("ITEM_TYPE")?.let {
            ItemType.of(it)
        } ?: ItemType.UNKNOWN,
        setting.getString("MATERIAL")?.let {
            Material.valueOf(it)
        } ?: throw IllegalArgumentException("Setting doesn't contain Material."),
        setting.getInt("MODEL"),
        TranslateUtil.get("FORMAT.NAME", null, setting),
        TranslateUtil.getList("FORMAT.LORE", null, setting)
    )
}